styx-common/src/main/java/com/spotify/styx/util/ShardedCounter.java

```diff
@@ -217,11 +251,65 @@ private CounterSnapshot refreshCounterSnapshot(String counterId) {
    * <p>Augments the transaction with operations to persist the given limit in Datastore. So long as
    * there has been no preceding successful updateLimit operation, no limit is applied in
    * updateCounter operations on this counter.
+   *
+   * <p>It spawns other transactions if needed to do redistribution of shards in a specific case:
+   * if newLimit < oldLimit and newLimit >= currentCounterUsage
+   *
+   * <p>Redistribution will be done before the limit is updated within the transaction as parameter, but
+   * that will still be a valid distribution for the old limit, even if the new limit update fails.
    */
-  public void updateLimit(DatastoreReaderWriter transaction, String counterId, long limit) {
+  public void updateLimit(DatastoreReaderWriter transaction, String counterId, long newLimit) {
     final Key limitKey = datastore.newKeyFactory().setKind(KIND_COUNTER_LIMIT).newKey(counterId);
 
-    transaction.put(Entity.newBuilder(limitKey).set(PROPERTY_LIMIT, limit).build());
+    final Entity limitEntity = transaction.get(limitKey);
+    final Long oldLimit =
+        limitEntity == null ? DEFAULT_MAX_COUNTER_LIMIT : limitEntity.getLong(PROPERTY_LIMIT);
+    transaction.put(Entity.newBuilder(limitKey).set(PROPERTY_LIMIT, newLimit).build());
+    redistributeShardsIfNeeded(counterId, newLimit, oldLimit);
+  }
+
+  /**
+   * Redistribution of shard values for the given counter and limits.
+   * Redistribution should be done before the limit update happens.
```

Comment:
"Should" or "must"? What are the requirements and guarantees this implementation is about to offer?

I'm seeing what seems to be not a 100% guarantee here - the redistribution is not in the same transaction as the limit update, so there is a race condition - any shard can become updated between those operations, by another instance, so the edge case is not fully averted.

Reply:
Not sure how it should be phrased best. Ideally it should be called only when the `limit` is updated (which here is the case at L268 https://github.com/spotify/styx/pull/396/files#diff-33665282cd4186e101219fbbe4c1faddR268). A `limit` is a completely separate entity, so the transactions doing the redistribution of the counter shards should not conflict with the transaction provided for the `limit` update.

That being said, there *could* be issues if the counter is in use/being updated while the redistribution is ongoing, but only if the counter surpasses the new `limit` that's being set. I can see that case causing an ongoing loop until the usage decreases below the new `limit`, but that's a case that can be addressed and fixed.